Explain why `non_snake_case` is skipped for binary crates and cleanup tests

jieyouxu opened the pull request:
- Explain `non_snake_case` lint is skipped for bin crate names because binaries are not intended to be distributed or consumed like library crates (#45127).
- Coalesce the bunch of tests into a single one but with revisions, which is easier to compare the differences for `non_snake_case` behavior with respect to crate types.

Follow-up to #121749 with some more comments and test cleanup.

cc @saethlin who bumped into one of the tests and was confused why it was `only-x86_64-unknown-linux-gnu`.

try-job: dist-i586-gnu-i586-i686-musl

jieyouxu committed:
Explain why non_snake_case is skipped for binary crates

jieyouxu committed:
Coalesce `non_snake_case` crate test set with revisions

jieyouxu commented:
@bors rollup=iffy (this might bounce due to unsupported crate types for specific targets)

Zalathar commented:
I notice that this also gets rid of two extra `bin` tests that set the crate type and crate name in a variety of ways.

I don't object to that, but I just want to double-check that it was intentional.

Zalathar commented:
Nah, if there's no reason to think the extra variants will be useful then it seems fine to get rid of them.

petrochenkov commented on tests/ui/lint/non-snake-case/lint-non-snake-case-crate.rs:
```diff
@@ -0,0 +1,26 @@
+//!
+//! See <https://github.com/rust-lang/rust/issues/45127>.
+
+//@ revisions: bin_ cdylib_ dylib_ lib_ proc_macro_ rlib_ staticlib_
```
```suggestion
//@ revisions: bin cdylib dylib lib proc_macro rlib staticlib
```
Why can't the versions without underscores be used for revisions?

jieyouxu replied:
E.g. `proc_macro` is a built-in cfg

```
error: unexpected `--cfg proc_macro` flag
   |
   = note: config `proc_macro` is only supposed to be controlled by `--crate-type proc-macro`
   = note: manually setting a built-in cfg can and does create incoherent behaviors
   = note: `#[deny(explicit_builtin_cfgs_in_flags)]` on by default
```

I just added a trailing `_` to all of them to be consistent.

jieyouxu replied:
@rustbot ready

petrochenkov commented:
r=me with https://github.com/rust-lang/rust/pull/130599#discussion_r1768903792 answered.
@rustbot author

petrochenkov commented:
@bors r+

bors commented:
:pushpin: Commit aebc28eaca07030147d826c4260bb66b9506af31 has been approved by `petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov aebc28eaca07030147d826c4260bb66b9506af31 -->
<!-- homu: {"type":"Approved","sha":"aebc28eaca07030147d826c4260bb66b9506af31","approver":"petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit aebc28eaca07030147d826c4260bb66b9506af31 with merge 5050837e308d1dfbdf782ebbbe4abc7ebf85a23e...
<!-- homu: {"type":"BuildStarted","head_sha":"aebc28eaca07030147d826c4260bb66b9506af31","merge_sha":"5050837e308d1dfbdf782ebbbe4abc7ebf85a23e"} -->

rust-log-analyzer commented:
The job **`dist-i586-gnu-i586-i686-musl`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/10969474160/job/30462166271) [(plain)](https://github.com/rust-lang-ci/rust/commit/5050837e308d1dfbdf782ebbbe4abc7ebf85a23e/checks/30462166271/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
test [ui] tests/ui/lint/non-local-defs/module_as_local.rs ... ok
test [ui] tests/ui/lint/non-local-defs/suggest-moving-inner.rs ... ok
test [ui] tests/ui/lint/non-local-defs/macro_rules.rs ... ok
test [ui] tests/ui/lint/non-local-defs/trait-solver-overflow-123573.rs ... ok
test [ui] tests/ui/lint/non-snake-case/lint-non-snake-case-crate.rs#cdylib_ ... ignored, ignored when the target is i686-unknown-linux-musl ((cdylib is not supported))
test [ui] tests/ui/lint/non-snake-case/lint-non-snake-case-crate.rs#bin_ ... ok
test [ui] tests/ui/lint/non-local-defs/weird-exprs.rs ... ok
test [ui] tests/ui/lint/non-snake-case/lint-non-snake-case-crate.rs#dylib_ ... FAILED
test [ui] tests/ui/lint/non-snake-case/lint-non-snake-case-crate.rs#lib_ ... ok
---
test [ui] tests/ui/wrong-hashset-issue-42918.rs ... ok

failures:

---- [ui] tests/ui/lint/non-snake-case/lint-non-snake-case-crate.rs#dylib_ stdout ----

error in revision `dylib_`: ui test compiled successfully!
status: exit status: 0
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lint/non-snake-case/lint-non-snake-case-crate.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=i686-unknown-linux-musl" "--cfg" "dylib_" "--check-cfg" "cfg(FALSE,bin_,cdylib_,dylib_,lib_,proc_macro_,rlib_,staticlib_)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/non-snake-case/lint-non-snake-case-crate.dylib_" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/i686-unknown-linux-musl/native/rust-test-helpers" "-Clinker=/musl-i686/bin/musl-gcc" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/non-snake-case/lint-non-snake-case-crate.dylib_/auxiliary" "--crate-type=dylib"
--- stderr -------------------------------
warning: dropping unsupported crate type `dylib` for target `i686-unknown-linux-musl`

warning: 1 warning emitted
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/10969474160/job/30462912083)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/10969474160/job/30462912083","builder_name":"checks-actions"} -->

jieyouxu commented:
@bors try

jieyouxu commented on tests/ui/lint/non-snake-case/lint-non-snake-case-crate.rs:
```diff
@@ -0,0 +1,29 @@
+
+// But should fire on non-binary crates.
+
+//@[cdylib_] ignore-musl (dylibs are not supported)
+//@[dylib_] ignore-musl (dylibs are not supported)
```
Changed the ignores to become `ignore-musl` because I think musl targets doesn't support dylibs in general.

bors commented:
:hourglass: Trying commit 9ae1fb4329eaed6e75d20da67b30bf86c5700481 with merge 883506d9cd1bcb4c442d6823eba69952f7f92be9...
<!-- homu: {"type":"TryBuildStarted","head_sha":"9ae1fb4329eaed6e75d20da67b30bf86c5700481","merge_sha":"883506d9cd1bcb4c442d6823eba69952f7f92be9"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/10970112460/job/30464582570)
Build commit: 883506d9cd1bcb4c442d6823eba69952f7f92be9 (`883506d9cd1bcb4c442d6823eba69952f7f92be9`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/10970112460/job/30464582570"},"merge_sha":"883506d9cd1bcb4c442d6823eba69952f7f92be9"} -->

jieyouxu commented:
Changes since last review:
- Fix test failure by changing to `ignore-musl` for both `cdylib` and `dylib` (musl AFAIK doesn't support dylibs in general)

@rustbot ready

petrochenkov commented:
@bors r+

bors commented:
:pushpin: Commit 9ae1fb4329eaed6e75d20da67b30bf86c5700481 has been approved by `petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov 9ae1fb4329eaed6e75d20da67b30bf86c5700481 -->
<!-- homu: {"type":"Approved","sha":"9ae1fb4329eaed6e75d20da67b30bf86c5700481","approver":"petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 9ae1fb4329eaed6e75d20da67b30bf86c5700481 with merge f48c99a0041bb82a750b34cf2570e0e73d6d801c...
<!-- homu: {"type":"BuildStarted","head_sha":"9ae1fb4329eaed6e75d20da67b30bf86c5700481","merge_sha":"f48c99a0041bb82a750b34cf2570e0e73d6d801c"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/10972133634/job/30470043732)
Approved by: petrochenkov
Pushing f48c99a0041bb82a750b34cf2570e0e73d6d801c to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"petrochenkov","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/10972133634/job/30470043732"},"merge_sha":"f48c99a0041bb82a750b34cf2570e0e73d6d801c"} -->

rust-timer commented:
Finished benchmarking commit (f48c99a0041bb82a750b34cf2570e0e73d6d801c): [comparison URL](https://perf.rust-lang.org/compare.html?start=74fd001cdae0321144a20133f2216ea8a97da476&end=f48c99a0041bb82a750b34cf2570e0e73d6d801c&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=74fd001cdae0321144a20133f2216ea8a97da476&end=f48c99a0041bb82a750b34cf2570e0e73d6d801c&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=74fd001cdae0321144a20133f2216ea8a97da476&end=f48c99a0041bb82a750b34cf2570e0e73d6d801c&stat=max-rss)
<details>
<summary>Results (primary -2.5%, secondary 2.0%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 2.0%  | [2.0%, 2.0%]   | 1     |
| Improvements ✅ <br /> (primary)   | -2.5% | [-2.5%, -2.5%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -2.5% | [-2.5%, -2.5%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=74fd001cdae0321144a20133f2216ea8a97da476&end=f48c99a0041bb82a750b34cf2570e0e73d6d801c&stat=cycles:u)
<details>
<summary>Results (secondary 1.8%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 1.8% | [1.8%, 1.8%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=74fd001cdae0321144a20133f2216ea8a97da476&end=f48c99a0041bb82a750b34cf2570e0e73d6d801c&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 769.225s -> 768.677s (-0.07%)
**Artifact size**: 341.35 MiB -> 341.36 MiB (0.00%)
